### ISSUE TITLE
Permission API for DApps

### ISSUE THREAD

**Issue by hayeah** [MetaMask/metamask-extension#3703]
hi, we are the https://keymesh.io team.

One major UX problem with DApp that depends on MetaMask is that each action requires an explicit approval from the user.

It would be nice to have a permission request API, like the "grant access to camera" modal in mobile apps. The permissions can be granted for specific low-value, low-risk, and high-frequency contract method calls.

For keymesh in particular, we could ask the users to grant the right (once, 30 days, or forever) to call high-frequency methods:

'broadcast'
'publishMessage'

**Comment by danfinlay** [comment]
I have explored the notion that there may be cases where a app shouldn't need user confirmation, here:
https://github.com/MetaMask/metamask-extension/issues/2532#issuecomment-341801971

In my proposal, I suggested that keys could be created with specific terms associated. In your proposal, a certain Dapp could get limited permissions regarding a certain key.

It's a tough issue. I agree user confirmation is an obstacle, but it's also one of the best security guarantees that we have at the moment.

One problem with this proposal, is that since it refers to on-chain transactions, and not just signatures, this means a user is agreeing to an unknown amount of gas fees. We cannot predict what the gas market will look like in the future (it sometimes seems like magic to even recommend for the present), and so a user consenting to send transactions that a dapp suggests in the future could mean agreeing to high gas fees.

I guess you could make an "up to gas cost" parameter for the permission.

There definitely could be a lot of caveats for something like this, adding any logic that bypasses user confirmation makes for some very dangerous code to write. I partly think that contract accounts could address solve these features by providing on-chain limited permissions to act on a user's behalf for some types of transactions.

Those are my initial thoughts, I agree that long term users will want ways for their smart identities to act on their behalves without confirmation, but this will need to be addressed with the highest security, it would take some very careful specifying before I'd be comfortable with a feature like this.

**Comment by hayeah** [comment]
Thanks for the thoughtful response @danfinlay. Cautiousness is a must. To address your concern with "agreeing to an unknown amount of gas/fees/values", my thinking is rather simple... 

1. When signing up a service, DApp can request for a budget of $10, $20, $30, $50 USD. The permission-less calls would drain from this budget. To users this like a pre-charge credit-cards on a per DApp basis, and they can revoke the budget at anytime.
2. As per above, the permissions would be granted on ABI method signatures. The budget is only used to pay for gas.
3. Any method call where payable value is non-zero MUST go through user confirmation.

This design is tailored to DApps as they exist today, and doesn't aim to be general like #2532

**Comment by danfinlay** [comment]
I think the simple way to enable this in the short term is to generate an ethereum account in the page, and ask the user to fund it. This allows the user to limit their exposure, but allows the app to submit transactions on their behalf. If you don't want to affect the user's holdings, you could use `WETH` and have them give your app an `allowance` for that ether.

My point partly being that if the user is handing over "control" to the app, that control can also be handed over on-chain. It may not need to be at the signer level.

This may be a fine feature in the longer term, but I think we have a lot of more basic usability issues to work out before we start adding things like this.

**Comment by danfinlay** [comment]
I was just looking for an old issue and came up on this one. I was just about to share this proposal, and I think you might be interested here too, it's a pretty spec'd out version of what you requested!:
https://github.com/danfinlay/EIPs/blob/WalletPermissions/EIPS/eip-PermissionsSystem.md

**Comment by hayeah** [comment]
@danfinlay is the intention for everything to be pre-approved at sign-in time? I could leave some comments at the actual EIP...

**Comment by Robbiekruszynski** [comment]
Hi @hayeah 

Since it looks there are a few tickets / conversations revolving around this that Dan mentioned above I'm going to go ahead and close this ticket.
